# Statistical Methods in Pandas - Lab

## Introduction

In this lab you'll get some hands-on experience using some of the key summary statistics methods in Pandas.

## Objectives
You will be able to:

- Calculate summary statistics for a series and DataFrame
- Use the `.apply()` or `.applymap()` methods to apply a function to a pandas series or DataFrame  


## Getting Started

For this lab, we'll be working with a dataset containing information on various lego datasets. You will find this dataset in the file `'lego_sets.csv'`.   

In the cell below:

- Import Pandas and set the standard alias of `pd`
- Import the `'lego_sets.csv'` dataset
- Display the first five rows of the DataFrame to get a feel for what we'll be working with

In [3]:
# Import pandas
import pandas as pd

# Import the 'lego_sets.csv' dataset
df = pd.read_csv('lego_sets.csv', engine='python', on_bad_lines='skip')

# Print the first five rows of DataFrame
df.head()

ages  list_price  num_reviews  piece_count  play_star_rating  \
0  6-12       29.99          2.0        277.0               4.0   
1  6-12       19.99          2.0        168.0               4.0   
2  6-12       12.99         11.0         74.0               4.3   
3   12+       99.99         23.0       1032.0               3.6   
4   12+       79.99         14.0        744.0               3.2   

                                           prod_desc  prod_id  \
0  Catapult into action and take back the eggs fr...  75823.0   
1  Launch a flying attack and rescue the eggs fro...  75822.0   
2  Chase the piggy with lightning-fast Chuck and ...  75821.0   
3  Explore the architecture of the United States ...  21030.0   
4  Recreate the Solomon R. Guggenheim Museum® wit...  21035.0   

                                      prod_long_desc review_difficulty  \
0  Use the staircase catapult to launch Red into ...           Average   
1  Pilot Pig has taken off from Bird Island with ...              Easy   
2  Pitch speedy bird Chuck against the Piggy Car....              Easy   
3  Discover the architectural secrets of the icon...           Average   
4  Discover the architectural secrets of Frank Ll...       Challenging   

                         set_name  star_rating    theme_name  val_star_rating  \
0           Bird Island Egg Heist          4.5  Angry Birds™              4.0   
1              Piggy Plane Attack          5.0  Angry Birds™              4.0   
2                Piggy Car Escape          4.3  Angry Birds™              4.1   
3  United States Capitol Building          4.6  Architecture              4.3   
4   Solomon R. Guggenheim Museum®          4.6  Architecture              4.1   

  country  
0      US  
1      US  
2      US  
3      US  
4      US

## Getting DataFrame-Level Statistics

We'll begin by getting some overall summary statistics on the dataset. There are two ways we'll get this information -- `.info()` and `.describe()`.

The `.info()` method provides us metadata on the DataFrame itself. This allows us to answer questions such as:

* What data type does each column contain?
* How many rows are in my dataset?
* How many total non-missing values does each column contain?
* How much memory does the DataFrame take up?

In the cell below, call our DataFrame's `.info()` method.

In [4]:
# Call the .info() method
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8975 entries, 0 to 8974
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ages               8975 non-null   object 
 1   list_price         8975 non-null   float64
 2   num_reviews        7804 non-null   float64
 3   piece_count        8975 non-null   float64
 4   play_star_rating   7691 non-null   float64
 5   prod_desc          8704 non-null   object 
 6   prod_id            8975 non-null   float64
 7   prod_long_desc     8975 non-null   object 
 8   review_difficulty  7488 non-null   object 
 9   set_name           8975 non-null   object 
 10  star_rating        7804 non-null   float64
 11  theme_name         8972 non-null   object 
 12  val_star_rating    7677 non-null   float64
 13  country            8975 non-null   object 
dtypes: float64(7), object(7)
memory usage: 981.8+ KB


#### Interpreting the Results

Read the output above, and then answer the following questions:

- How many total rows are in this DataFrame?  
- How many columns contain numeric data?
- How many contain categorical data?
- Identify at least 3 columns that contain missing values.

Write your answer below this line:
________________________________________________________________________________________________________________________________

How many total rows are in this DataFrame? **8975**

How many columns contain numeric data? **7**

How many contain categorical data? **7**

Identify at least 3 columns that contain missing values. num_reviews , player_star_rating and prod_desc

Whereas `.info()` provides statistics about the DataFrame itself, `.describe()` returns output containing basic summary statistics about the data contained with the DataFrame.  

In the cell below, call the DataFrame's `.describe()` method.

In [5]:
# Call the .describe() method
df.describe()

list_price  num_reviews  piece_count  play_star_rating       prod_id  \
count  8975.000000  7804.000000  8975.000000       7691.000000  8.975000e+03   
mean     62.908689    16.624936   487.464067          4.333975  5.974649e+04   
std      89.286695    35.956150   817.571644          0.653664  1.621033e+05   
min       2.272400     1.000000     1.000000          1.000000  6.300000e+02   
25%      19.492200     2.000000    95.000000          4.000000  2.103600e+04   
50%      35.598000     6.000000   213.000000          4.500000  4.206900e+04   
75%      68.307800    13.000000   521.000000          4.800000  7.092200e+04   
max    1097.939000   367.000000  7541.000000          5.000000  2.000431e+06   

       star_rating  val_star_rating  
count  7804.000000      7677.000000  
mean      4.513813         4.228748  
std       0.518143         0.660426  
min       1.800000         1.000000  
25%       4.300000         4.000000  
50%       4.700000         4.300000  
75%       5.000000         4.700000  
max       5.000000         5.000000

#### Interpreting the Results

The output contains descriptive statistics corresponding to the columns. Use these to answer the following questions:

- How much is the standard deviation for `piece count`?
- How many pieces are in the largest lego set?
- How many in the smallest lego set? What is the median `val_star_rating`?

________________________________________________________________________________________________________________________________

How much is the standard deviation for piece count? **817.571644**

How many pieces are in the largest lego set? **7541.000000**

How many in the smallest lego set? **1.000000**

What is the median val_star_rating? **4.300000**


## Getting Summary Statistics

Pandas also allows us to easily compute individual summary statistics using built-in methods.  Next, we'll get some practice using these methods.

In the cell below, compute the median value of the `star_rating` column.

In [6]:
# Calculate the median of the star_rating column
df['star_rating'].median()

4.7

Next, get a count of the total number of unique values in `play_star_rating`.

In [7]:
# Print the number of unique values in play_star_rating
df['play_star_rating'].nunique()

30

Now, compute the standard deviation of the `list_price` column.

In [8]:
# Calculate the standard deviation of the list_price column
df['list_price'].std()

89.28669540346516

If we bought every single lego set in this dataset, how many pieces would we have?  

> **Note**: If you truly want to answer this accurately, and are up for the challenge, remove duplicate lego-set entries before summing the pieces. That is, many of the lego sets are listed multiple times in the dataset above, depending on the country where it is being sold and other unique parameters. If you're stuck, just practice calculating the total number of pieces in the dataset for now.

In [9]:
# Total number of pieces across all unique Lego sets
df_unique = df.drop_duplicates(subset=['prod_id'])
total_pieces = df_unique['piece_count'].sum()

print(f"Total number of LEGO pieces (unique sets only): {int(total_pieces):,}")

Total number of LEGO pieces (unique sets only): 319,071


Now, let's try getting the value for the 90% quantile for all numerical columns.  Do this in the cell below.

In [11]:
# Get the 90% quantile for all numerical columns
numerical_df = df.select_dtypes(include=['float64', 'int64'])
display(numerical_df.quantile(0.9))

list_price            134.139
num_reviews            38.000
piece_count          1053.000
play_star_rating        5.000
prod_id             75531.000
star_rating             5.000
val_star_rating         5.000
Name: 0.9, dtype: float64

## Getting Summary Statistics on Categorical Data

For obvious reasons, most of the methods we've used so far only work with numerical data -- there's no way to calculate the standard deviation of a column containing string values. However, there are some things that we can discover about columns containing categorical data.

In the cell below, print the unique values contained within the `review_difficulty` column.

In [12]:
# Print the unique values in the review_difficulty column
df['review_difficulty'].unique()

array(['Average', 'Easy', 'Challenging', 'Very Easy', nan,
       'Very Challenging'], dtype=object)

Now, let's get the `value_counts()` for this column, to see how common each is.

In [13]:
# Get the value_counts() of the review_difficulty column
df['review_difficulty'].value_counts()

review_difficulty
Easy                3116
Average             2756
Very Easy            849
Challenging          761
Very Challenging       6
Name: count, dtype: int64

As you can see, these provide us quick and easy ways to get information on columns containing categorical information.  


## Using `.applymap()`

When working with pandas DataFrames, we can quickly compute functions on the data contained by using the `.applymap()` method and passing in a lambda function.

For instance, we can use `applymap()` to return a version of the DataFrame where every value has been converted to a string.

In the cell below:

* Call the DataFrame's `.applymap()` method and pass in `lambda x: str(x)`  
* Call the new `string_df` object's `.info()` method to confirm that everything has been cast to a string

In [14]:
# Call the .applymap() method
string_df = df.applymap(lambda x: str(x))

# Call the .info() method
string_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8975 entries, 0 to 8974
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ages               8975 non-null   object
 1   list_price         8975 non-null   object
 2   num_reviews        8975 non-null   object
 3   piece_count        8975 non-null   object
 4   play_star_rating   8975 non-null   object
 5   prod_desc          8975 non-null   object
 6   prod_id            8975 non-null   object
 7   prod_long_desc     8975 non-null   object
 8   review_difficulty  8975 non-null   object
 9   set_name           8975 non-null   object
 10  star_rating        8975 non-null   object
 11  theme_name         8975 non-null   object
 12  val_star_rating    8975 non-null   object
 13  country            8975 non-null   object
dtypes: object(14)
memory usage: 981.8+ KB


/tmp/ipython-input-14-1211924104.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_df = df.applymap(lambda x: str(x))


Note that everything -- even the `NaN` values, have been cast to a string in the example above.

Note that for Pandas Series objects (such as a single column in a DataFrame), we can do the same thing using the `.apply()` method.  

This is just one example of how we can quickly compute custom functions on our DataFrame.

## Summary

In this lab, we learned how to:

* Use the `df.describe()` and `df.info()` summary statistics methods
* Use built-in Pandas methods for calculating summary statistics
* Apply a function to every element in a DataFrame